File: internal/service/crawler/remote_ok.go
package crawler

import (
	"context"
	"fmt"
	"jd-matcher/utility"
	"strings"

	"github.com/anaskhan96/soup"
	"github.com/gogf/gf/v2/frame/g"
	"github.com/gogf/gf/v2/text/gstr"
	"github.com/gogf/gf/v2/util/gconv"
)

const REMOTE_OK_BASE_URL = "https://remoteok.com"

func GetRemoteOkJobs(ctx context.Context, jobTypes, locations []string, offset int) ([]CommonJob, error) {

	var reqUrl = REMOTE_OK_BASE_URL + "/?action=get_jobs&offset=" + gconv.String(offset)
	resp, err := utility.GetContent(ctx, reqUrl)
	if err != nil {
		return nil, err
	}

	jobs := parseRemoteOkMainPageJobs(ctx, resp)

	return jobs, nil
}

func parseRemoteOkMainPageJobs(ctx context.Context, htmlStr string) (jobs []CommonJob) {
	defer func(ctx context.Context) {
		if rec := recover(); rec != nil {
			g.Log().Line().Error(ctx, fmt.Sprintf("parse remote ok job list failed: %s", rec))
		}
	}(ctx)

	// if htmlStr start with <tr, then surround it with <table></table>
	htmlStr = gstr.Trim(htmlStr)
	if strings.HasPrefix(htmlStr, "<tr") {
		htmlStr = "<table>" + htmlStr + "</table>"
	}

	docs := soup.HTMLParse(htmlStr)
	if docs.Error != nil {
		panic(docs.Error)
	}
	jobExpends := docs.FindAll("tr", "class", "expand")

	for _, jobExpend := range jobExpends {
		var (
			jobTitle       string
			jobUrl         string
			jobDescription string
		)
		if jobExpend.Find("input", "class", "share-job-copy-paste").Pointer == nil {
			continue
		}
		if jobExpend.Find("div", "class", "html").Pointer != nil {
			jobDescription = jobExpend.Find("div", "class", "html").FullText()
		} else if jobExpend.Find("div", "class", "markdown").Pointer != nil {
			jobDescription = jobExpend.Find("div", "class", "markdown").FullText()
		}

		job := CommonJob{
			Description: gstr.Trim(jobDescription),
		}

		dataId := jobExpend.Attrs()["data-id"]
		if dataId != "" {
			// get job url
			jobHeader := docs.Find("tr", "class", "job-"+dataId)
			jobUrl = jobHeader.Attrs()["data-url"]
			job.Url = REMOTE_OK_BASE_URL + jobUrl

			// get job title
			companyPosistionTrDoc := jobHeader.Find("td", "class", "company_and_position")
			if companyPosistionTrDoc.Pointer == nil {
				continue
			}
			jobTitle = companyPosistionTrDoc.Find("h2").FullText()
			job.Title = gstr.Trim(jobTitle)

			// get job location and salary
			jobLocationSalaryDivDocs := jobHeader.FindAll("div", "class", "location")
			locationList := []string{}
			for i := 0; i < len(jobLocationSalaryDivDocs); i++ {
				if i == len(jobLocationSalaryDivDocs)-1 {
					job.Salary = gstr.Trim(jobLocationSalaryDivDocs[i].Text())
				} else {
					locationList = append(locationList, gstr.Trim(jobLocationSalaryDivDocs[i].Text()))
				}
			}

			if len(locationList) > 0 {
				job.Location = strings.Join(locationList, ",")
			}

			// get job tags
			if jobHeader.Pointer != nil {
				jobTags := jobHeader.Find("td", "class", "tags")
				jobTagsH3 := jobTags.FindAll("h3")

				for _, jobTagH3 := range jobTagsH3 {
					job.Tags = append(job.Tags, gstr.TrimAll(jobTagH3.Text()))
				}
			}
		}

		jobs = append(jobs, job)
	}

	return jobs
}

File: src/lib/crawler/remote_ok.ts
import * as cheerio from "cheerio";
import type { CommonJob } from "../types.js";

const BASE = "https://remoteok.com";

export async function getRemoteOkJobs(offset = 1): Promise<CommonJob[]> {
  const resp = await fetch(`${BASE}/?action=get_jobs&offset=${offset}`);
  const html = await resp.text();
  return parseRemoteOkMainPageJobs(html);
}

export function parseRemoteOkMainPageJobs(htmlStr: string): CommonJob[] {
  const jobs: CommonJob[] = [];
  let html = htmlStr.trim();
  if (html.startsWith("<tr")) html = "<table>" + html + "</table>";

  const $ = cheerio.load(html);
  $("tr.expand").each((_i, el) => {
    try {
      const $el = $(el);
      const dataId = $el.attr("data-id");
      if (!dataId) return;

      const header = $(`tr.job-${dataId}`);
      if (!header.length) return;

      const jobUrl = header.attr("data-url") ?? "";
      const title = header.find("td.company_and_position h2").text().trim();

      const htmlDiv = $el.find("div.html");
      const mdDiv = $el.find("div.markdown");
      const desc = htmlDiv.length ? htmlDiv.text() : mdDiv.length ? mdDiv.text() : "";

      const locDivs = header.find("div.location");
      const salaryDiv = header.find("div.salary");

      const locs: string[] = [];
      locDivs.each((_i, e) => {
        const t = $(e).text().trim();
        if (!t) return;
        if (/^⏰\s/.test(t)) return; // employment type (Part time, Full time, Contractor)
        if (t.includes("Upgrade to Premium")) return; // "💰 Upgrade to Premium to see salary"
        locs.push(t);
      });

      let salary = salaryDiv.length ? salaryDiv.text().trim() : "";
      if (!salary) {
        const premium = header.find("div.location").filter((_i, e) => $(e).text().includes("Upgrade to Premium"));
        if (premium.length) salary = premium.text().trim();
      }

      const time = header.find("time").attr("datetime") ?? "";
      const tags: string[] = [];
      header.find("td.tags h3").each((_j, e) => { tags.push($(e).text().trim()); });

      jobs.push({
        title,
        url: BASE + jobUrl,
        description: desc.trim(),
        tags,
        location: locs.join(","),
        salary,
        updateTime: time,
      });
    } catch {
      /* skip malformed */
    }
  });
  return jobs;
}
